{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1c2f7044-f3eb-4386-8b3f-6f7fce5c7ae8",
   "metadata": {
    "tags": []
   },
   "source": [
    "for markdown basic syntax see [here](https://www.markdownguide.org/basic-syntax/)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45448eea-8d13-4aa2-8e60-6d8bee1a56d7",
   "metadata": {},
   "source": [
    "# Nucleotide to amino acid sequence conversion, protein alignment and protein function prediction"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71f0da6a-0a71-4572-b4dc-d042d11216b0",
   "metadata": {},
   "source": [
    "**After calculating the desired SNPs for each sample**, the coding sequences need to be extracted, converted to protein sequences and these then need to be aligned in order to ascertain which SNPs are causing amino acid changes. VCF cleanup, SNP categorization and read support percentage calculation should be done beforehand.  \n",
    "To do this, a **GFF3 file with reference annotation** is needed as well as the reference **genome.fasta** and a **consensus.fasta** for each sample at each desired SNP cutoff level."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "975c1408-0820-4089-b8d7-203dca7f1a33",
   "metadata": {},
   "source": [
    "**Tools**: gffread, blastp, local bash and ubuntu tools."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f593d3bf-1888-4d86-aa08-1096b1d120a7",
   "metadata": {},
   "source": [
    "#### Inputs:\n",
    "Don't put the final \"/\" on paths.  \n",
    "Variables with \"name\" should contain just the name of the file and variables without it should contain the complete path to the file excluding the format.  \n",
    "Arrays **cds_list** and **gene_csv** are redundant but should be kept as such."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81f51fed-32cd-4d09-a608-b250446174ba",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#desired folder for the output of this notebook\n",
    "analysis_dir='/home/fuberlin/Documents/Runs/Illumina/20220421_MDV_HSV1_BAC/MDV'\n",
    "mkdir \"$analysis_dir\"\n",
    "cd \"$analysis_dir\"/\n",
    "\n",
    "#if output folder does not exist one will be created\n",
    "output_dir='/home/fuberlin/Documents/Runs/Illumina/20220421_MDV_HSV1_BAC/MDV/Protein_Prediction/output' \n",
    "mkdir \"$output_dir\"\n",
    "\n",
    "#complete path to a ref FASTA file, and separate variable for the name of the reference\n",
    "reference='/home/fuberlin/Documents/Runs/Illumina/20220421_MDV_HSV1_BAC/MDV/Protein_Prediction/MDV_ref'\n",
    "refname='MDV_ref'\n",
    "\n",
    "#annotation is a gff3 file that can be extracted directly from the ncbi accession page\n",
    "annotation='/home/fuberlin/Documents/Runs/Illumina/20220421_MDV_HSV1_BAC/MDV/Protein_Prediction/MDV_ref'\n",
    "\n",
    "#codingsequences is a txt file that can be extracted directly from the ncbi accession page, separate variable for the name of the codingsequences file\n",
    "codingsequences='/home/fuberlin/Documents/Runs/Illumina/20220421_MDV_HSV1_BAC/MDV/Protein_Prediction/MDV_coding_sequences.txt'\n",
    "csname='MDV_coding_sequences.txt'\n",
    "\n",
    "#proteinsequences is a txt file that can be extracted directly from the ncbi accession page, separate variable for the name of the proteinsequences file\n",
    "proteinsequences='/home/fuberlin/Documents/Runs/Illumina/20220421_MDV_HSV1_BAC/MDV/Protein_Prediction/MDV_protein_sequences.txt'\n",
    "psname='MDV_protein_sequences.txt'\n",
    "\n",
    "#samples is a list of the sample codes generated in sequencing, ending in _SXX\n",
    "samples=(CEF-WT-PLX-2_S9 CEF-WT-PX-1_S1 CEF-WT-PX-2_S3 CEF-WT-PXXX-1_S5 CEF-WT-PXXX-2_S7 CEF-YS-PLX-2_S10 CEF-YS-PX-1_S2 CEF-YS-PX-2_S4 CEF-YS-PXXX-1_S6 CEF-YS-PXXX-2_S8 T7-WT-PXII_S11 T7-WT-PXX_S12)\n",
    "\n",
    "\n",
    "#snp_dir is the folder where the SNP calculations were made (should contain 1 folder per sample of standard SNP outputs as per the following script: Variant_Calling_6modes_MN08022022.sh)\n",
    "snp_dir='/home/fuberlin/Documents/Runs/Illumina/20220421_MDV_HSV1_BAC/MDV/SNP_Prediction'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c07ec5cc-1dea-494b-8458-94ce6c0ac42e",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Reference (gene IDs and locations)\n",
    "Getting a table of gene IDs and locations for labelling of SNPs later on."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8484245f-c90f-4e2d-bf89-f3a90b8a8877",
   "metadata": {},
   "outputs": [],
   "source": [
    "grep -e \">\" \"$codingsequences\" > cs1.txt\n",
    "sed -e 's/^.*gene/gene/g' -e 's/gene\\=//g' -e 's/\\].*\\[location\\=/\\ /g' -e 's/\\.\\..*\\.\\./\\.\\./g' -e 's/\\.\\./\\ /g' -e 's/complement(//g' -e 's/join(//g' -e 's/)\\]//g' -e 's/\\[gbkey\\=CDS\\]//g' -e 's/\\.\\./\\ /g' -e 's/\\]//g' -e 's/\\ /,/' -e 's/\\ /,/' -e 's/)//g' -e 's/>.*\\=/>/g' -e 's/\\,/\\ /g' -e 's/>//g' -e 's/\\ /\\,/g' cs1.txt > cs.csv\n",
    "echo \"CSV generated\"\n",
    "gene_csv=( $(tail -n +1 cs.csv | cut -d ',' -f1) )\n",
    "start_csv=( $(tail -n +1 cs.csv | cut -d ',' -f2) )\n",
    "end_csv=( $(tail -n +1 cs.csv | cut -d ',' -f3) )\n",
    "#Report\n",
    "echo \"Gene ID and location arrays generated.\"\n",
    "echo \"\"${gene_csv[@]}\"\"\n",
    "echo \"\"${start_csv[@]}\"\"\n",
    "echo \"\"${end_csv[@]}\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6951cbdd-4d49-425a-8233-c4f33427dc2f",
   "metadata": {},
   "source": [
    "### Reference (nt to aa)\n",
    "Generating the protein repertoire from the reference, to align all the proteins from the sample consensus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a43d816d-bf53-4e58-8163-14056ba9e55c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#extraction of protein information using gffread\n",
    "#gffread -C \"$annotation\".gff3 -g \"$reference\".fasta -y proteins.fasta\n",
    "sed -e 's/^.*gene/>gene/g' -e 's/gene\\=//g' -e 's/\\].*\\[location\\=/\\ /g' -e 's/\\.\\..*\\.\\./\\.\\./g' -e 's/\\.\\./\\ /g' -e 's/complement(//g' -e 's/join(//g' -e 's/)\\]//g' -e 's/\\[gbkey\\=CDS\\]//g' -e 's/\\.\\./\\ /g' -e 's/\\]//g' -e 's/\\ /,/' -e 's/\\ /,/' -e 's/)//g' -e 's/>.*\\=/>/g' -e 's/\\,.*//g' -e 's/>/>gene\\-/g' \"$proteinsequences\" > proteins.fasta\n",
    "awk '/^>/ {$0=$0 \"_ref\"}1' proteins.fasta>protein.fasta\n",
    "#create folder for protein fastas and move into it\n",
    "mkdir \"$output_dir\"/\"$refname\"/\n",
    "#move file to protein fastas folder\n",
    "mv protein.fasta proteins.fasta \"$output_dir\"/\"$refname\"/\n",
    "cd \"$output_dir\"/\"$refname\"/\n",
    "cat protein.fasta | awk '{\n",
    "        if (substr($0, 1, 1)==\">\") {filename=(substr($0,2) \".fasta\")}\n",
    "        print $0 >> filename\n",
    "        close(filename)\n",
    "}'\n",
    "#Report\n",
    "echo \"Reference protein sequences generated.\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c4c6899-9d79-4f95-988b-3429cdd29555",
   "metadata": {},
   "source": [
    "### NT to AA (samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca61a22d-a953-4b39-9fae-9a2b4c4fd912",
   "metadata": {},
   "source": [
    "The headers of the consensus files must match the headers of the fasta reference and GFF3 file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02e38a55-1660-441c-8fa7-10f9eda1ccfd",
   "metadata": {},
   "source": [
    "#### Importing the consensus files and renaming them\n",
    "This may not be needed anymore if the VC scripts are updated the consensus files are renamed at the point of their creation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5180b0a5-297f-4763-ab31-185dfaaf80bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir \"$analysis_dir\"/csensus/\n",
    "cd \"$snp_dir\"\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "#for 50% SNP read support\n",
    "cd \"$snp_dir\"/\"$s\"/Variant_Calling_0.5_SNPs/\n",
    "cp *.fa \"$analysis_dir\"/csensus/\n",
    "cd \"$analysis_dir\"/csensus/\n",
    "rename 's/_freebayes_0.5_SNPs.fa/_50.fasta/g' *.fa\n",
    "\n",
    "#for 10% SNP read support\n",
    "cd \"$snp_dir\"/\"$s\"/Variant_Calling_0.1_SNPs/\n",
    "cp *.fa \"$analysis_dir\"/csensus/\n",
    "cd \"$analysis_dir\"/csensus/\n",
    "rename 's/_freebayes_0.1_SNPs.fa/_10.fasta/g' *.fa\n",
    "\n",
    "#for 5% SNP read support\n",
    "cd \"$snp_dir\"/\"$s\"/Variant_Calling_0.05_SNPs/\n",
    "cp *.fa \"$analysis_dir\"/csensus/\n",
    "cd \"$analysis_dir\"/csensus/\n",
    "rename 's/_freebayes_0.05_SNPs.fa/_5.fasta/g' *.fa\n",
    "\n",
    "#for 1% SNP read support\n",
    "cd \"$snp_dir\"/\"$s\"/Variant_Calling_0.01_SNPs/\n",
    "cp *.fa \"$analysis_dir\"/csensus/\n",
    "cd \"$analysis_dir\"/csensus/\n",
    "rename 's/_freebayes_0.01_SNPs.fa/_1.fasta/g' *.fa\n",
    "\n",
    "#Report\n",
    "echo \"Consensus file for \"$s\" re-formatted.\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "059e945a-a568-4297-b82b-835131cc2553",
   "metadata": {},
   "source": [
    "For HSV1 Pol Mutant data, the reference was generated by using sequencing data generated in house to calculate SNPs and apply them to the Strain F reference. This was done because the stocks used for this work should already be a few nucleotides away from the published Strain F reference.  \n",
    "  \n",
    "**ENTER FASTA HEADER HERE**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f9eea63-608f-4666-a0ac-8db8e108b6ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "#to match the headers of the consensus files with the header of the reference, shouldn't be needed in other runs\n",
    "for fasta in $(find \"$analysis_dir\"/csensus/ -name \"*.fasta\")\n",
    "do\n",
    "    new_fasta=$(basename $fasta .fasta).new.fasta\n",
    "    sed -e 's/>.*/>EF523390.1/g' $fasta > $new_fasta\n",
    "done\n",
    "mkdir ./corrected/\n",
    "mv *.new.fasta ./corrected/\n",
    "rm *.fasta\n",
    "cd ./corrected/\n",
    "mv *.fasta \"$analysis_dir\"/csensus/\n",
    "cd \"$analysis_dir\"/csensus/\n",
    "rm -r ./corrected/\n",
    "rename 's/.new.fasta/.fasta/g' *.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1466b6c4-cbf5-4bce-9901-a32e1f630901",
   "metadata": {},
   "source": [
    "#### Getting the protein sequences\n",
    "Generating the CDS/protein repertoire from each sample at 50% and 10% cutoff. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5b29953-9381-467a-9e5c-d430a2af6d22",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "cd \"$analysis_dir\"/\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "\n",
    "#for 50% SNP read support\n",
    "gffread \"$annotation\".gff3 -g \"$analysis_dir\"/csensus/consensus_\"$s\"_50.fasta -y pt_\"$s\"_50.fasta\n",
    "#add sample to the end of each title line \">\"\n",
    "awk '/^>/ {$0=$0 \"_'\"$s\"'\"}1' pt_\"$s\"_50.fasta>protein_\"$s\"_50.fasta\n",
    "mkdir \"$output_dir\"/\"$s\"_50/\n",
    "mv pt_\"$s\"_50.fasta protein_\"$s\"_50.fasta \"$output_dir\"/\"$s\"_50/\n",
    "cd \"$output_dir\"/\"$s\"_50/\n",
    "#separating the multifasta into individual fastas named after the lines with \">\"\n",
    "cat protein_\"$s\"_50.fasta | awk '{\n",
    "        if (substr($0, 1, 1)==\">\") {filename=(substr($0,2) \".fasta\")}\n",
    "        print $0 >> filename\n",
    "        close(filename)\n",
    "}'\n",
    "\n",
    "#for 10% SNP read support\n",
    "gffread \"$annotation\".gff3 -g \"$analysis_dir\"/csensus/consensus_\"$s\"_10.fasta -y pt_\"$s\"_10.fasta\n",
    "#add sample to the end of each title line \">\"\n",
    "awk '/^>/ {$0=$0 \"_'\"$s\"'\"}1' pt_\"$s\"_10.fasta>protein_\"$s\"_10.fasta\n",
    "mkdir \"$output_dir\"/\"$s\"_10/\n",
    "mv pt_\"$s\"_10.fasta protein_\"$s\"_10.fasta \"$output_dir\"/\"$s\"_10/\n",
    "cd \"$output_dir\"/\"$s\"_10/\n",
    "#separating the multifasta into individual fastas named after the lines with \">\"\n",
    "cat protein_\"$s\"_10.fasta | awk '{\n",
    "        if (substr($0, 1, 1)==\">\") {filename=(substr($0,2) \".fasta\")}\n",
    "        print $0 >> filename\n",
    "        close(filename)\n",
    "}'\n",
    "\n",
    "#for 5% SNP read support\n",
    "gffread \"$annotation\".gff3 -g \"$analysis_dir\"/csensus/consensus_\"$s\"_5.fasta -y pt_\"$s\"_5.fasta\n",
    "#add sample to the end of each title line \">\"\n",
    "awk '/^>/ {$0=$0 \"_'\"$s\"'\"}1' pt_\"$s\"_5.fasta>protein_\"$s\"_5.fasta\n",
    "mkdir \"$output_dir\"/\"$s\"_5/\n",
    "mv pt_\"$s\"_5.fasta protein_\"$s\"_5.fasta \"$output_dir\"/\"$s\"_5/\n",
    "cd \"$output_dir\"/\"$s\"_5/\n",
    "#separating the multifasta into individual fastas named after the lines with \">\"\n",
    "cat protein_\"$s\"_5.fasta | awk '{\n",
    "        if (substr($0, 1, 1)==\">\") {filename=(substr($0,2) \".fasta\")}\n",
    "        print $0 >> filename\n",
    "        close(filename)\n",
    "}'\n",
    "\n",
    "#for 1% SNP read support\n",
    "gffread \"$annotation\".gff3 -g \"$analysis_dir\"/csensus/consensus_\"$s\"_1.fasta -y pt_\"$s\"_1.fasta\n",
    "#add sample to the end of each title line \">\"\n",
    "awk '/^>/ {$0=$0 \"_'\"$s\"'\"}1' pt_\"$s\"_1.fasta>protein_\"$s\"_1.fasta\n",
    "mkdir \"$output_dir\"/\"$s\"_1/\n",
    "mv pt_\"$s\"_1.fasta protein_\"$s\"_1.fasta \"$output_dir\"/\"$s\"_1/\n",
    "cd \"$output_dir\"/\"$s\"_1/\n",
    "#separating the multifasta into individual fastas named after the lines with \">\"\n",
    "cat protein_\"$s\"_1.fasta | awk '{\n",
    "        if (substr($0, 1, 1)==\">\") {filename=(substr($0,2) \".fasta\")}\n",
    "        print $0 >> filename\n",
    "        close(filename)\n",
    "}'\n",
    "\n",
    "#Report\n",
    "echo \"\"$s\" protein sequences generated.\"\n",
    "\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0136d61c-d443-4e7f-b0c0-cbb6f3a946ed",
   "metadata": {},
   "source": [
    "### Aligning protein sequences\n",
    "Getting the corresponding proteins from the reference and sample repertoires and aligning them. To do this, we need to iterate through the list of samples but also through a list of proteins that is generated using the full protein repertoire of the reference before it is separated into ssFASTA files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d67f1cf1-fd13-42a0-a5fe-bb8c417560ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "#grabbing a list of all CDS/proteins available from the seq\n",
    "cd \"$output_dir\"/\"$refname\"/\n",
    "grep -e \">\" proteins.fasta >list1.txt\n",
    "awk 'sub(/^>/, \"\")' list1.txt >list2.txt \n",
    "rm list1.txt\n",
    "readarray -t cds_list < list2.txt \n",
    "echo \"${cds_list[@]}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7b1db0d-1c04-43c8-a2fa-c6e8048c4bfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Protein alignments for 50% SNP support sequences\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "    mkdir \"$output_dir\"/\"$s\"_50/alignments/\n",
    "    cd \"$output_dir\"/\"$s\"_50/alignments/\n",
    "    for cds in \"${cds_list[@]}\"\n",
    "    do\n",
    "    #When not provided, the default value is: 'qaccver saccver pident length mismatch gapopen qstart qend sstart send evalue bitscore', which is equivalent to the keyword 'std'\n",
    "        sed -i -e '2,$s/\\.//g' \"$output_dir\"/\"$s\"_50/\"$cds\"_\"$s\".fasta\n",
    "        blastp -query \"$output_dir\"/\"$s\"_50/\"$cds\"_\"$s\".fasta -subject \"$output_dir\"/\"$refname\"/\"$cds\"_ref.fasta -out a_\"$cds\"_\"$s\".txt -outfmt 10\n",
    "        tr , '\\n' < a_\"$cds\"_\"$s\".txt > tmp_\"$cds\"_\"$s\".txt\n",
    "        rm a_\"$cds\"_\"$s\".txt\n",
    "        readarray -t a < tmp_\"$cds\"_\"$s\".txt\n",
    "        if [[ \"${a[4]}\" == 0 ]]\n",
    "        then\n",
    "            rm tmp_\"$cds\"_\"$s\".txt\n",
    "        elif [[ \"${a[4]}\" > 0 ]]\n",
    "        then\n",
    "            blastp -query \"$output_dir\"/\"$s\"_50/\"$cds\"_\"$s\".fasta -subject \"$output_dir\"/\"$refname\"/\"$cds\"_ref.fasta -out alignment_\"$cds\"_\"$s\"_50.txt -outfmt 5\n",
    "            rm tmp_\"$cds\"_\"$s\".txt\n",
    "        else\n",
    "           echo \"error on \"$s\": \"$cds\".\"\n",
    "        fi\n",
    "    done\n",
    "done\n",
    "#Report\n",
    "echo \"Protein alignments for sequences at 50% SNP read support done.\"\n",
    "\n",
    "#Protein alignment for 10% SNP support sequences\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "    mkdir \"$output_dir\"/\"$s\"_10/alignments/\n",
    "    cd \"$output_dir\"/\"$s\"_10/alignments/\n",
    "    for cds in \"${cds_list[@]}\"\n",
    "    do\n",
    "    #When not provided, the default value is: 'qaccver saccver pident length mismatch gapopen qstart qend sstart send evalue bitscore', which is equivalent to the keyword 'std'\n",
    "        sed -i -e '2,$s/\\.//g' \"$output_dir\"/\"$s\"_10/\"$cds\"_\"$s\".fasta\n",
    "        blastp -query \"$output_dir\"/\"$s\"_10/\"$cds\"_\"$s\".fasta -subject \"$output_dir\"/\"$refname\"/\"$cds\"_ref.fasta -out a_\"$cds\"_\"$s\".txt -outfmt 10\n",
    "        tr , '\\n' < a_\"$cds\"_\"$s\".txt > tmp_\"$cds\"_\"$s\".txt\n",
    "        rm a_\"$cds\"_\"$s\".txt\n",
    "        readarray -t a < tmp_\"$cds\"_\"$s\".txt\n",
    "        if [[ \"${a[4]}\" == 0 ]]\n",
    "        then\n",
    "            rm tmp_\"$cds\"_\"$s\".txt\n",
    "        elif [[ \"${a[4]}\" > 0 ]]\n",
    "        then\n",
    "            blastp -query \"$output_dir\"/\"$s\"_10/\"$cds\"_\"$s\".fasta -subject \"$output_dir\"/\"$refname\"/\"$cds\"_ref.fasta -out alignment_\"$cds\"_\"$s\"_10.txt -outfmt 5\n",
    "            rm tmp_\"$cds\"_\"$s\".txt\n",
    "        else\n",
    "            echo \"error on \"$s\": \"$cds\".\"\n",
    "        fi\n",
    "    done\n",
    "done\n",
    "#Report\n",
    "echo \"Protein alignments for sequences at 10% SNP read support done.\"\n",
    "\n",
    "#Protein alignments for 5% SNP support sequences\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "    mkdir \"$output_dir\"/\"$s\"_5/alignments/\n",
    "    cd \"$output_dir\"/\"$s\"_5/alignments/\n",
    "    for cds in \"${cds_list[@]}\"\n",
    "    do\n",
    "    #When not provided, the default value is: 'qaccver saccver pident length mismatch gapopen qstart qend sstart send evalue bitscore', which is equivalent to the keyword 'std'\n",
    "        sed -i -e '2,$s/\\.//g' \"$output_dir\"/\"$s\"_5/\"$cds\"_\"$s\".fasta\n",
    "        blastp -query \"$output_dir\"/\"$s\"_5/\"$cds\"_\"$s\".fasta -subject \"$output_dir\"/\"$refname\"/\"$cds\"_ref.fasta -out a_\"$cds\"_\"$s\".txt -outfmt 10\n",
    "        tr , '\\n' < a_\"$cds\"_\"$s\".txt > tmp_\"$cds\"_\"$s\".txt\n",
    "        rm a_\"$cds\"_\"$s\".txt\n",
    "        readarray -t a < tmp_\"$cds\"_\"$s\".txt\n",
    "        if [[ \"${a[4]}\" == 0 ]]\n",
    "        then\n",
    "            rm tmp_\"$cds\"_\"$s\".txt\n",
    "        elif [[ \"${a[4]}\" > 0 ]]\n",
    "        then\n",
    "            blastp -query \"$output_dir\"/\"$s\"_5/\"$cds\"_\"$s\".fasta -subject \"$output_dir\"/\"$refname\"/\"$cds\"_ref.fasta -out alignment_\"$cds\"_\"$s\"_5.txt -outfmt 5\n",
    "            rm tmp_\"$cds\"_\"$s\".txt\n",
    "        else\n",
    "           echo \"error on \"$s\": \"$cds\".\"\n",
    "        fi\n",
    "    done\n",
    "done\n",
    "#Report\n",
    "echo \"Protein alignments for sequences at 5% SNP read support done.\"\n",
    "\n",
    "#Protein alignment for 1% SNP support sequences\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "    mkdir \"$output_dir\"/\"$s\"_1/alignments/\n",
    "    cd \"$output_dir\"/\"$s\"_1/alignments/\n",
    "    for cds in \"${cds_list[@]}\"\n",
    "    do\n",
    "    #When not provided, the default value is: 'qaccver saccver pident length mismatch gapopen qstart qend sstart send evalue bitscore', which is equivalent to the keyword 'std'\n",
    "        sed -i -e '2,$s/\\.//g' \"$output_dir\"/\"$s\"_1/\"$cds\"_\"$s\".fasta\n",
    "        blastp -query \"$output_dir\"/\"$s\"_1/\"$cds\"_\"$s\".fasta -subject \"$output_dir\"/\"$refname\"/\"$cds\"_ref.fasta -out a_\"$cds\"_\"$s\".txt -outfmt 10\n",
    "        tr , '\\n' < a_\"$cds\"_\"$s\".txt > tmp_\"$cds\"_\"$s\".txt\n",
    "        rm a_\"$cds\"_\"$s\".txt\n",
    "        readarray -t a < tmp_\"$cds\"_\"$s\".txt\n",
    "        if [[ \"${a[4]}\" == 0 ]]\n",
    "        then\n",
    "            rm tmp_\"$cds\"_\"$s\".txt\n",
    "        elif [[ \"${a[4]}\" > 0 ]]\n",
    "        then\n",
    "            blastp -query \"$output_dir\"/\"$s\"_1/\"$cds\"_\"$s\".fasta -subject \"$output_dir\"/\"$refname\"/\"$cds\"_ref.fasta -out alignment_\"$cds\"_\"$s\"_1.txt -outfmt 5\n",
    "            rm tmp_\"$cds\"_\"$s\".txt\n",
    "        else\n",
    "            echo \"error on \"$s\": \"$cds\".\"\n",
    "        fi\n",
    "    done\n",
    "done\n",
    "#Report\n",
    "echo \"Protein alignments for sequences at 1% SNP read support done.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6836635d-cf0e-4a11-9f81-5dac0844932a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generating CSV files for AA changes in each sample at 50% SNP read support\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "    cd \"$output_dir\"/\"$s\"_50/alignments/\n",
    "    mkdir \"$output_dir\"/\"$s\"_50/working/\n",
    "    mkdir \"$output_dir\"/\"$s\"_50/aa_changes/\n",
    "    echo \"Gene, Start, End, Possible SNP Start (Direct), Possible SNP End (Direct), Possible SNP Start (Complement), Possible SNP End (Complement), AA Change, PROVEAN Score, PROVEAN Prediction\" > \"$output_dir\"/\"$s\"_50/aa_changes/\"$s\"_aa_change_50.csv\n",
    "    counter=0\n",
    "    for filename in *\n",
    "    do\n",
    "    counter=$((counter+1))\n",
    "    cp \"$filename\" \"$output_dir\"/\"$s\"_50/working/\n",
    "    cd \"$output_dir\"/\"$s\"_50/working/\n",
    "    sed '1,48d' \"$filename\" > a_\"$counter\".txt\n",
    "    sed '4,22d' a_\"$counter\".txt > b_\"$counter\".txt\n",
    "    sed -e 's/\\<Hsp_qseq\\>//g' -e 's;\\<\\/\\-Hsp\\_qseq\\>;;g' -e 's/\\<Hsp_hseq\\>//g' -e 's/\\<\\/\\-Hsp\\_hseq\\>//g' -e 's/\\<Hsp_midline\\>//g' -e 's/\\<\\/\\-Hsp_midline\\>//g' b_\"$counter\".txt > c_\"$counter\".txt\n",
    "    cat c_\"$counter\".txt | tr -d \"[:blank:]\" > d_\"$counter\".txt\n",
    "    sed 's/^.\\{,2\\}//' d_\"$counter\".txt > e_\"$counter\".txt\n",
    "    sed 's/...$//' e_\"$counter\".txt > f_\"$counter\".txt\n",
    "    sed -n 1p f_\"$counter\".txt > text1_\"$counter\".txt\n",
    "    sed -n 2p f_\"$counter\".txt > text2_\"$counter\".txt\n",
    "    sed 's/./&\\n/g' text1_\"$counter\".txt > t1_\"$counter\".txt\n",
    "    sed 's/./&\\n/g' text2_\"$counter\".txt > t2_\"$counter\".txt\n",
    "    readarray -t query < t1_\"$counter\".txt\n",
    "    readarray -t subj < t2_\"$counter\".txt\n",
    "    #substituting AAs that are the same in reference and query arrays for dots\n",
    "        for a in \"${!query[@]}\" \n",
    "        do\n",
    "            if [[ \"${query[$a]}\" == \"${subj[$a]}\" ]]\n",
    "            then\n",
    "                query[$a]=\".\"\n",
    "            fi\n",
    "        done\n",
    "    #printing info about unique (non-dot) AAs in the query array into aa_changes.txt in the format: alignmentfile: AAinref|index+1|AAinquery, in which index+1 represents the AA position\n",
    "        for b in \"${!query[@]}\"\n",
    "        do\n",
    "            if [[ ${query[$b]} != \".\" ]]\n",
    "            then\n",
    "                i=$((b+1))\n",
    "                for gene in \"${gene_csv[@]}\"\n",
    "                do\n",
    "                    if [[ \"$filename\" =~ .*\"$gene\"_.* ]]\n",
    "                    then\n",
    "                        for j in \"${!gene_csv[@]}\"\n",
    "                        do\n",
    "                            if [[ \"${gene_csv[$j]}\" == \"$gene\" ]];\n",
    "                            then\n",
    "                            st_fr=$((start_csv[$j]+(\"$i\"*3)-3))\n",
    "                            end_fr=$((start_csv[$j]+(\"$i\"*3)+3))\n",
    "                            st_cmp=$((end_csv[$j]-(\"$i\"*3)-3))\n",
    "                            end_cmp=$((end_csv[$j]-(\"$i\"*3)+3))\n",
    "                            echo \"\"$gene\",\"${start_csv[$j]}\",\"${end_csv[$j]}\",\"$st_fr\",\"$end_fr\",\"$st_cmp\",\"$end_cmp\",\"${subj[$b]}\"\"$i\"\"${query[$b]}\"\" >> \"$output_dir\"/\"$s\"_50/aa_changes/\"$s\"_aa_change_50.csv\n",
    "                            #Creating the var files necessay for Provean analysis in the next steps\n",
    "                            cd \"$output_dir\"/\"$s\"_50/aa_changes/\n",
    "                            echo \"\" > \"$gene\".var\n",
    "                            echo \"\"${subj[$b]}\"\"$i\"\"${query[$b]}\"\" >> \"$gene\".var\n",
    "                            cd \"$output_dir\"/\"$s\"_50/working/\n",
    "                            fi\n",
    "                        done\n",
    "                    fi\n",
    "                done\n",
    "            fi\n",
    "        done\n",
    "    #change back to original folder so that the loop won't break\n",
    "    cd \"$output_dir\"/\"$s\"_50/alignments/\n",
    "    done\n",
    "rm -r \"$output_dir\"/\"$s\"_50/working/\n",
    "#Report\n",
    "echo \"Amino acid change CSV file generated for sample \"$s\" at 50% SNP read support.\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0da03ae6-f09b-452c-8ed3-f3cc93320479",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generating CSV files for AA changes in each sample at 10% SNP read support\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "    cd \"$output_dir\"/\"$s\"_10/alignments/\n",
    "    mkdir \"$output_dir\"/\"$s\"_10/working/\n",
    "    mkdir \"$output_dir\"/\"$s\"_10/aa_changes/\n",
    "    echo \"Gene, Start, End, Possible SNP Start (Direct), Possible SNP End (Direct), Possible SNP Start (Complement), Possible SNP End (Complement), AA Change, PROVEAN Score, PROVEAN Prediction\" > \"$output_dir\"/\"$s\"_10/aa_changes/\"$s\"_aa_change_10.csv\n",
    "    counter=0\n",
    "    for filename in *\n",
    "    do\n",
    "    counter=$((counter+1))\n",
    "    cp \"$filename\" \"$output_dir\"/\"$s\"_10/working/\n",
    "    cd \"$output_dir\"/\"$s\"_10/working/\n",
    "    sed '1,48d' \"$filename\" > a_\"$counter\".txt\n",
    "    sed '4,22d' a_\"$counter\".txt > b_\"$counter\".txt\n",
    "    sed -e 's/\\<Hsp_qseq\\>//g' -e 's;\\<\\/\\-Hsp\\_qseq\\>;;g' -e 's/\\<Hsp_hseq\\>//g' -e 's/\\<\\/\\-Hsp\\_hseq\\>//g' -e 's/\\<Hsp_midline\\>//g' -e 's/\\<\\/\\-Hsp_midline\\>//g' b_\"$counter\".txt > c_\"$counter\".txt\n",
    "    cat c_\"$counter\".txt | tr -d \"[:blank:]\" > d_\"$counter\".txt\n",
    "    sed 's/^.\\{,2\\}//' d_\"$counter\".txt > e_\"$counter\".txt\n",
    "    sed 's/...$//' e_\"$counter\".txt > f_\"$counter\".txt\n",
    "    sed -n 1p f_\"$counter\".txt > text1_\"$counter\".txt\n",
    "    sed -n 2p f_\"$counter\".txt > text2_\"$counter\".txt\n",
    "    sed 's/./&\\n/g' text1_\"$counter\".txt > t1_\"$counter\".txt\n",
    "    sed 's/./&\\n/g' text2_\"$counter\".txt > t2_\"$counter\".txt\n",
    "    readarray -t query < t1_\"$counter\".txt\n",
    "    readarray -t subj < t2_\"$counter\".txt\n",
    "    #substituting AAs that are the same in reference and query arrays for dots\n",
    "        for a in \"${!query[@]}\" \n",
    "        do\n",
    "            if [[ \"${query[$a]}\" == \"${subj[$a]}\" ]]\n",
    "            then\n",
    "                query[$a]=\".\"\n",
    "            fi\n",
    "        done\n",
    "    #printing info about unique (non-dot) AAs in the query array into aa_changes.txt in the format: alignmentfile: AAinref|index+1|AAinquery, in which index+1 represents the AA position\n",
    "        for b in \"${!query[@]}\"\n",
    "        do\n",
    "            if [[ ${query[$b]} != \".\" ]]\n",
    "            then\n",
    "                i=$((b+1))\n",
    "                for gene in \"${gene_csv[@]}\"\n",
    "                do\n",
    "                    if [[ \"$filename\" =~ .*\"$gene\"_.* ]]\n",
    "                    then\n",
    "                        for j in \"${!gene_csv[@]}\"\n",
    "                        do\n",
    "                            if [[ \"${gene_csv[$j]}\" == \"$gene\" ]];\n",
    "                            then\n",
    "                            st_fr=$((start_csv[$j]+(\"$i\"*3)-3))\n",
    "                            end_fr=$((start_csv[$j]+(\"$i\"*3)+3))\n",
    "                            st_cmp=$((end_csv[$j]-(\"$i\"*3)-3))\n",
    "                            end_cmp=$((end_csv[$j]-(\"$i\"*3)+3))\n",
    "                            echo \"\"$gene\",\"${start_csv[$j]}\",\"${end_csv[$j]}\",\"$st_fr\",\"$end_fr\",\"$st_cmp\",\"$end_cmp\",\"${subj[$b]}\"\"$i\"\"${query[$b]}\"\" >> \"$output_dir\"/\"$s\"_10/aa_changes/\"$s\"_aa_change_10.csv\n",
    "                            #Creating the var files necessay for Provean analysis in the next steps\n",
    "                            cd \"$output_dir\"/\"$s\"_10/aa_changes/\n",
    "                            echo \"\" > \"$gene\".var\n",
    "                            echo \"\"${subj[$b]}\"\"$i\"\"${query[$b]}\"\" >> \"$gene\".var\n",
    "                            cd \"$output_dir\"/\"$s\"_10/working/\n",
    "                            fi\n",
    "                        done\n",
    "                    fi\n",
    "                done\n",
    "            fi\n",
    "        done\n",
    "    #change back to original folder so that the loop won't break\n",
    "    cd \"$output_dir\"/\"$s\"_10/alignments/\n",
    "    done\n",
    "rm -r \"$output_dir\"/\"$s\"_10/working/\n",
    "#Report\n",
    "echo \"Amino acid change CSV file generated for sample \"$s\" at 10% SNP read support.\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f252fd77-c993-4716-884d-07c9d08908a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generating CSV files for AA changes in each sample at 5% SNP read support\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "    cd \"$output_dir\"/\"$s\"_5/alignments/\n",
    "    mkdir \"$output_dir\"/\"$s\"_5/working/\n",
    "    mkdir \"$output_dir\"/\"$s\"_5/aa_changes/\n",
    "    echo \"Gene, Start, End, Possible SNP Start (Direct), Possible SNP End (Direct), Possible SNP Start (Complement), Possible SNP End (Complement), AA Change, PROVEAN Score, PROVEAN Prediction\" > \"$output_dir\"/\"$s\"_5/aa_changes/\"$s\"_aa_change_5.csv\n",
    "    counter=0\n",
    "    for filename in *\n",
    "    do\n",
    "    counter=$((counter+1))\n",
    "    cp \"$filename\" \"$output_dir\"/\"$s\"_5/working/\n",
    "    cd \"$output_dir\"/\"$s\"_5/working/\n",
    "    sed '1,48d' \"$filename\" > a_\"$counter\".txt\n",
    "    sed '4,22d' a_\"$counter\".txt > b_\"$counter\".txt\n",
    "    sed -e 's/\\<Hsp_qseq\\>//g' -e 's;\\<\\/\\-Hsp\\_qseq\\>;;g' -e 's/\\<Hsp_hseq\\>//g' -e 's/\\<\\/\\-Hsp\\_hseq\\>//g' -e 's/\\<Hsp_midline\\>//g' -e 's/\\<\\/\\-Hsp_midline\\>//g' b_\"$counter\".txt > c_\"$counter\".txt\n",
    "    cat c_\"$counter\".txt | tr -d \"[:blank:]\" > d_\"$counter\".txt\n",
    "    sed 's/^.\\{,2\\}//' d_\"$counter\".txt > e_\"$counter\".txt\n",
    "    sed 's/...$//' e_\"$counter\".txt > f_\"$counter\".txt\n",
    "    sed -n 1p f_\"$counter\".txt > text1_\"$counter\".txt\n",
    "    sed -n 2p f_\"$counter\".txt > text2_\"$counter\".txt\n",
    "    sed 's/./&\\n/g' text1_\"$counter\".txt > t1_\"$counter\".txt\n",
    "    sed 's/./&\\n/g' text2_\"$counter\".txt > t2_\"$counter\".txt\n",
    "    readarray -t query < t1_\"$counter\".txt\n",
    "    readarray -t subj < t2_\"$counter\".txt\n",
    "    #substituting AAs that are the same in reference and query arrays for dots\n",
    "        for a in \"${!query[@]}\" \n",
    "        do\n",
    "            if [[ \"${query[$a]}\" == \"${subj[$a]}\" ]]\n",
    "            then\n",
    "                query[$a]=\".\"\n",
    "            fi\n",
    "        done\n",
    "    #printing info about unique (non-dot) AAs in the query array into aa_changes.txt in the format: alignmentfile: AAinref|index+1|AAinquery, in which index+1 represents the AA position\n",
    "        for b in \"${!query[@]}\"\n",
    "        do\n",
    "            if [[ ${query[$b]} != \".\" ]]\n",
    "            then\n",
    "                i=$((b+1))\n",
    "                for gene in \"${gene_csv[@]}\"\n",
    "                do\n",
    "                    if [[ \"$filename\" =~ .*\"$gene\"_.* ]]\n",
    "                    then\n",
    "                        for j in \"${!gene_csv[@]}\"\n",
    "                        do\n",
    "                            if [[ \"${gene_csv[$j]}\" == \"$gene\" ]];\n",
    "                            then\n",
    "                            st_fr=$((start_csv[$j]+(\"$i\"*3)-3))\n",
    "                            end_fr=$((start_csv[$j]+(\"$i\"*3)+3))\n",
    "                            st_cmp=$((end_csv[$j]-(\"$i\"*3)-3))\n",
    "                            end_cmp=$((end_csv[$j]-(\"$i\"*3)+3))\n",
    "                            echo \"\"$gene\",\"${start_csv[$j]}\",\"${end_csv[$j]}\",\"$st_fr\",\"$end_fr\",\"$st_cmp\",\"$end_cmp\",\"${subj[$b]}\"\"$i\"\"${query[$b]}\"\" >> \"$output_dir\"/\"$s\"_5/aa_changes/\"$s\"_aa_change_5.csv\n",
    "                            #Creating the var files necessay for Provean analysis in the next steps\n",
    "                            cd \"$output_dir\"/\"$s\"_5/aa_changes/\n",
    "                            echo \"\" > \"$gene\".var\n",
    "                            echo \"\"${subj[$b]}\"\"$i\"\"${query[$b]}\"\" >> \"$gene\".var\n",
    "                            cd \"$output_dir\"/\"$s\"_5/working/\n",
    "                            fi\n",
    "                        done\n",
    "                    fi\n",
    "                done\n",
    "            fi\n",
    "        done\n",
    "    #change back to original folder so that the loop won't break\n",
    "    cd \"$output_dir\"/\"$s\"_5/alignments/\n",
    "    done\n",
    "rm -r \"$output_dir\"/\"$s\"_5/working/\n",
    "#Report\n",
    "echo \"Amino acid change CSV file generated for sample \"$s\" at 5% SNP read support.\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97cd710b-2fd9-41f0-9700-e7e1811e52db",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Generating CSV files for AA changes in each sample at 1% SNP read support\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "    cd \"$output_dir\"/\"$s\"_1/alignments/\n",
    "    mkdir \"$output_dir\"/\"$s\"_1/working/\n",
    "    mkdir \"$output_dir\"/\"$s\"_1/aa_changes/\n",
    "    echo \"Gene, Start, End, Possible SNP Start (Direct), Possible SNP End (Direct), Possible SNP Start (Complement), Possible SNP End (Complement), AA Change, PROVEAN Score, PROVEAN Prediction\" > \"$output_dir\"/\"$s\"_1/aa_changes/\"$s\"_aa_change_1.csv\n",
    "    counter=0\n",
    "    for filename in *\n",
    "    do\n",
    "    counter=$((counter+1))\n",
    "    cp \"$filename\" \"$output_dir\"/\"$s\"_1/working/\n",
    "    cd \"$output_dir\"/\"$s\"_1/working/\n",
    "    sed '1,48d' \"$filename\" > a_\"$counter\".txt\n",
    "    sed '4,22d' a_\"$counter\".txt > b_\"$counter\".txt\n",
    "    sed -e 's/\\<Hsp_qseq\\>//g' -e 's;\\<\\/\\-Hsp\\_qseq\\>;;g' -e 's/\\<Hsp_hseq\\>//g' -e 's/\\<\\/\\-Hsp\\_hseq\\>//g' -e 's/\\<Hsp_midline\\>//g' -e 's/\\<\\/\\-Hsp_midline\\>//g' b_\"$counter\".txt > c_\"$counter\".txt\n",
    "    cat c_\"$counter\".txt | tr -d \"[:blank:]\" > d_\"$counter\".txt\n",
    "    sed 's/^.\\{,2\\}//' d_\"$counter\".txt > e_\"$counter\".txt\n",
    "    sed 's/...$//' e_\"$counter\".txt > f_\"$counter\".txt\n",
    "    sed -n 1p f_\"$counter\".txt > text1_\"$counter\".txt\n",
    "    sed -n 2p f_\"$counter\".txt > text2_\"$counter\".txt\n",
    "    sed 's/./&\\n/g' text1_\"$counter\".txt > t1_\"$counter\".txt\n",
    "    sed 's/./&\\n/g' text2_\"$counter\".txt > t2_\"$counter\".txt\n",
    "    readarray -t query < t1_\"$counter\".txt\n",
    "    readarray -t subj < t2_\"$counter\".txt\n",
    "    #substituting AAs that are the same in reference and query arrays for dots\n",
    "        for a in \"${!query[@]}\" \n",
    "        do\n",
    "            if [[ \"${query[$a]}\" == \"${subj[$a]}\" ]]\n",
    "            then\n",
    "                query[$a]=\".\"\n",
    "            fi\n",
    "        done\n",
    "    #printing info about unique (non-dot) AAs in the query array into aa_changes.txt in the format: alignmentfile: AAinref|index+1|AAinquery, in which index+1 represents the AA position\n",
    "        for b in \"${!query[@]}\"\n",
    "        do\n",
    "            if [[ ${query[$b]} != \".\" ]]\n",
    "            then\n",
    "                i=$((b+1))\n",
    "                for gene in \"${gene_csv[@]}\"\n",
    "                do\n",
    "                    if [[ \"$filename\" =~ .*\"$gene\"_.* ]]\n",
    "                    then\n",
    "                        for j in \"${!gene_csv[@]}\"\n",
    "                        do\n",
    "                            if [[ \"${gene_csv[$j]}\" == \"$gene\" ]];\n",
    "                            then\n",
    "                            st_fr=$((start_csv[$j]+(\"$i\"*3)-3))\n",
    "                            end_fr=$((start_csv[$j]+(\"$i\"*3)+3))\n",
    "                            st_cmp=$((end_csv[$j]-(\"$i\"*3)-3))\n",
    "                            end_cmp=$((end_csv[$j]-(\"$i\"*3)+3))\n",
    "                            echo \"\"$gene\",\"${start_csv[$j]}\",\"${end_csv[$j]}\",\"$st_fr\",\"$end_fr\",\"$st_cmp\",\"$end_cmp\",\"${subj[$b]}\"\"$i\"\"${query[$b]}\"\" >> \"$output_dir\"/\"$s\"_1/aa_changes/\"$s\"_aa_change_1.csv\n",
    "                            #Creating the var files necessay for Provean analysis in the next steps\n",
    "                            cd \"$output_dir\"/\"$s\"_1/aa_changes/\n",
    "                            echo \"\" > \"$gene\".var\n",
    "                            echo \"\"${subj[$b]}\"\"$i\"\"${query[$b]}\"\" >> \"$gene\".var\n",
    "                            cd \"$output_dir\"/\"$s\"_1/working/\n",
    "                            fi\n",
    "                        done\n",
    "                    fi\n",
    "                done\n",
    "            fi\n",
    "        done\n",
    "    #change back to original folder so that the loop won't break\n",
    "    cd \"$output_dir\"/\"$s\"_1/alignments/\n",
    "    done\n",
    "rm -r \"$output_dir\"/\"$s\"_1/working/\n",
    "#Report\n",
    "echo \"Amino acid change CSV file generated for sample \"$s\" at 1% SNP read support.\"\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bd3ecc8-14ab-40b6-81ae-faede281767b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Housekeeping, removing all unecessary files\n",
    "cd \"$analysis_dir\"/\n",
    "mkdir results\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "mkdir -p results/\"$s\"/aa_changes/\n",
    "mv \"$output_dir\"/\"$s\"_50/aa_changes/\"$s\"_aa_change_50.csv \"$analysis_dir\"/results/\"$s\"/aa_changes/\n",
    "mv \"$output_dir\"/\"$s\"_10/aa_changes/\"$s\"_aa_change_10.csv \"$analysis_dir\"/results/\"$s\"/aa_changes/\n",
    "mv \"$output_dir\"/\"$s\"_5/aa_changes/\"$s\"_aa_change_5.csv \"$analysis_dir\"/results/\"$s\"/aa_changes/\n",
    "mv \"$output_dir\"/\"$s\"_1/aa_changes/\"$s\"_aa_change_1.csv \"$analysis_dir\"/results/\"$s\"/aa_changes/\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d3d1508-8e41-4d44-9811-e01c587837ab",
   "metadata": {},
   "source": [
    "## VCF File Formatting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfd9ca95-feaf-4107-8c1b-c919b541a7fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#VCF Formatting for 50% SNP read support\n",
    "for s in \"${samples[@]}\"\n",
    "do \n",
    "cd \"$snp_dir\"/\"$s\"/\"$s\"_Export/\n",
    "sed '1,60d' \"$snp_dir\"/\"$s\"/\"$s\"_Export/bwa_paired_\"$s\"_sorted_freebayes_0.5_SNPs.vcf > 1_50.vcf\n",
    "sed -e \"s/AB\\=.*:GL//g\" -e \"s/.\\/.\\://g\" -e \"s/\\:/\\,/1\" -e \"s/\\:.*//g\" -e \"s/\\s\\+/,/g\" -e \"s/\\,ID//g\" -e \"s/\\,FILTER\\,INFO\\,FORMAT\\,unknown/\\,DEPTH\\,REF\\_READS\\,ALT\\_READS\\,TYPE/g\" -e \"s/\\,\\.//g\" -e \"s/\\#CHROM\\,//g\" -e \"s/\\,/\\:/1\" -e \"s/.*\\://g\" -e \"s/REF/POS\\,REF/g\" 1_50.vcf > 2_50.vcf\n",
    "echo \"POS,REF,ALT,QUAL,DEPTH,REF_READS,ALT_READS,TYPE\" > 3_50.vcf\n",
    "    while IFS= read -r line; \n",
    "    do\n",
    "        if [[ \"$line\" =~ \",A,\" ]] && [[ \"$line\" =~ \",C,\" ]]\n",
    "        then\n",
    "           echo \"\"$line\",transversion\" >> 3_50.vcf\n",
    "        elif [[ \"$line\" =~ \",A,\" ]] && [[ \"$line\" =~ \",G,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transition\" >> 3_50.vcf\n",
    "        elif [[ \"$line\" =~ \",A,\" ]] && [[ \"$line\" =~ \",T,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transversion\" >> 3_50.vcf\n",
    "        elif [[ \"$line\" =~ \",C,\" ]] && [[ \"$line\" =~ \",G,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transversion\" >> 3_50.vcf\n",
    "        elif [[ \"$line\" =~ \",C,\" ]] && [[ \"$line\" =~ \",T,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transition\" >> 3_50.vcf\n",
    "        elif [[ \"$line\" =~ \",G,\" ]] && [[ \"$line\" =~ \",T,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transversion\" >> 3_50.vcf\n",
    "        else\n",
    "            echo \"\"$line\",substitution\" >> 3_50.vcf    \n",
    "        fi\n",
    "    done < 2_50.vcf\n",
    "sed '2d' 3_50.vcf  > 4_50.vcf\n",
    "sed -i '1d' 4_50.vcf\n",
    "echo \"POS,REF,ALT,QUAL,DEPTH,REF_READS,ALT_READS,TYPE,GENE\" > \"$s\"_SNPs_50.vcf\n",
    "    while IFS= read -r line; \n",
    "    do\n",
    "    pos=$(echo \"\"$line\"\" | sed  's/,.*$//');\n",
    "    nc_region='TRUE'\n",
    "        for c in \"${!gene_csv[@]}\"\n",
    "        do\n",
    "            if (( \"$pos\" >= \"${start_csv[$c]}\" && \"$pos\" <= \"${end_csv[$c]}\" ));\n",
    "            then\n",
    "                nc_region='FALSE'\n",
    "                echo \"\"$line\",\"\"${gene_csv[$c]}\"\",\" >> \"$s\"_SNPs_50.vcf;\n",
    "            fi\n",
    "        done\n",
    "        if [[ \"$nc_region\" == 'TRUE' ]];\n",
    "        then \n",
    "            echo \"\"$line\",-,\" >> \"$s\"_SNPs_50.vcf;\n",
    "        fi\n",
    "        nc_region='TRUE'\n",
    "    done < 4_50.vcf\n",
    "rm 1_50.vcf\n",
    "rm 2_50.vcf\n",
    "rm 3_50.vcf\n",
    "rm 4_50.vcf\n",
    "done\n",
    "\n",
    "#VCF Formatting for 10% SNP read support\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "cd \"$snp_dir\"/\"$s\"/\"$s\"_Export/\n",
    "sed '1,60d' \"$snp_dir\"/\"$s\"/\"$s\"_Export/bwa_paired_\"$s\"_sorted_freebayes_0.1_SNPs.vcf > 1_10.vcf\n",
    "sed -e \"s/AB\\=.*:GL//g\" -e \"s/.\\/.\\://g\" -e \"s/\\:/\\,/1\" -e \"s/\\:.*//g\" -e \"s/\\s\\+/,/g\" -e \"s/\\,ID//g\" -e \"s/\\,FILTER\\,INFO\\,FORMAT\\,unknown/\\,DEPTH\\,REF\\_READS\\,ALT\\_READS\\,TYPE/g\" -e \"s/\\,\\.//g\" -e \"s/\\#CHROM\\,//g\" -e \"s/\\,/\\:/1\" -e \"s/.*\\://g\" -e \"s/REF/POS\\,REF/g\" 1_10.vcf > 2_10.vcf\n",
    "echo \"POS,REF,ALT,QUAL,DEPTH,REF_READS,ALT_READS,TYPE,GENE\" > 3_10.vcf\n",
    "    while IFS= read -r line; \n",
    "    do\n",
    "        if [[ \"$line\" =~ \",A,\" ]] && [[ \"$line\" =~ \",C,\" ]]\n",
    "        then\n",
    "           echo \"\"$line\",transversion\" >> 3_10.vcf\n",
    "        elif [[ \"$line\" =~ \",A,\" ]] && [[ \"$line\" =~ \",G,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transition\" >> 3_10.vcf\n",
    "        elif [[ \"$line\" =~ \",A,\" ]] && [[ \"$line\" =~ \",T,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transversion\" >> 3_10.vcf\n",
    "        elif [[ \"$line\" =~ \",C,\" ]] && [[ \"$line\" =~ \",G,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transversion\" >> 3_10.vcf\n",
    "        elif [[ \"$line\" =~ \",C,\" ]] && [[ \"$line\" =~ \",T,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transition\" >> 3_10.vcf\n",
    "        elif [[ \"$line\" =~ \",G,\" ]] && [[ \"$line\" =~ \",T,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transversion\" >> 3_10.vcf\n",
    "        else\n",
    "            echo \"\"$line\",substitution\" >> 3_10.vcf   \n",
    "        fi\n",
    "    done < 2_10.vcf\n",
    "sed '2d' 3_10.vcf  > 4_10.vcf\n",
    "sed -i '1d' 4_10.vcf\n",
    "echo \"POS,REF,ALT,QUAL,DEPTH,REF_READS,ALT_READS,TYPE,GENE\" > \"$s\"_SNPs_10.vcf\n",
    "    while IFS= read -r line; \n",
    "    do\n",
    "    pos=$(echo \"\"$line\"\" | sed  's/,.*$//');\n",
    "    nc_region='TRUE'\n",
    "        for c in \"${!gene_csv[@]}\"\n",
    "        do\n",
    "            if (( \"$pos\" >= \"${start_csv[$c]}\" && \"$pos\" <= \"${end_csv[$c]}\" ));\n",
    "            then\n",
    "                nc_region='FALSE'\n",
    "                echo \"\"$line\",\"\"${gene_csv[$c]}\"\",\" >> \"$s\"_SNPs_10.vcf;\n",
    "            fi\n",
    "        done\n",
    "        if [[ \"$nc_region\" == 'TRUE' ]];\n",
    "        then \n",
    "            echo \"\"$line\",-,\" >> \"$s\"_SNPs_10.vcf;\n",
    "        fi\n",
    "        nc_region='TRUE'\n",
    "    done < 4_10.vcf\n",
    "rm 1_10.vcf\n",
    "rm 2_10.vcf\n",
    "rm 3_10.vcf\n",
    "rm 4_10.vcf\n",
    "done\n",
    "\n",
    "#VCF Formatting for 5% SNP read support\n",
    "for s in \"${samples[@]}\"\n",
    "do \n",
    "cd \"$snp_dir\"/\"$s\"/\"$s\"_Export/\n",
    "sed '1,60d' \"$snp_dir\"/\"$s\"/\"$s\"_Export/bwa_paired_\"$s\"_sorted_freebayes_0.05_SNPs.vcf > 1_5.vcf\n",
    "sed -e \"s/AB\\=.*:GL//g\" -e \"s/.\\/.\\://g\" -e \"s/\\:/\\,/1\" -e \"s/\\:.*//g\" -e \"s/\\s\\+/,/g\" -e \"s/\\,ID//g\" -e \"s/\\,FILTER\\,INFO\\,FORMAT\\,unknown/\\,DEPTH\\,REF\\_READS\\,ALT\\_READS\\,TYPE/g\" -e \"s/\\,\\.//g\" -e \"s/\\#CHROM\\,//g\" -e \"s/\\,/\\:/1\" -e \"s/.*\\://g\" -e \"s/REF/POS\\,REF/g\" 1_5.vcf > 2_5.vcf\n",
    "echo \"POS,REF,ALT,QUAL,DEPTH,REF_READS,ALT_READS,TYPE\" > 3_5.vcf\n",
    "    while IFS= read -r line; \n",
    "    do\n",
    "        if [[ \"$line\" =~ \",A,\" ]] && [[ \"$line\" =~ \",C,\" ]]\n",
    "        then\n",
    "           echo \"\"$line\",transversion\" >> 3_5.vcf\n",
    "        elif [[ \"$line\" =~ \",A,\" ]] && [[ \"$line\" =~ \",G,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transition\" >> 3_5.vcf\n",
    "        elif [[ \"$line\" =~ \",A,\" ]] && [[ \"$line\" =~ \",T,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transversion\" >> 3_5.vcf\n",
    "        elif [[ \"$line\" =~ \",C,\" ]] && [[ \"$line\" =~ \",G,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transversion\" >> 3_5.vcf\n",
    "        elif [[ \"$line\" =~ \",C,\" ]] && [[ \"$line\" =~ \",T,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transition\" >> 3_5.vcf\n",
    "        elif [[ \"$line\" =~ \",G,\" ]] && [[ \"$line\" =~ \",T,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transversion\" >> 3_5.vcf\n",
    "        else\n",
    "            echo \"\"$line\",substitution\" >> 3_5.vcf    \n",
    "        fi\n",
    "    done < 2_5.vcf\n",
    "sed '2d' 3_5.vcf  > 4_5.vcf\n",
    "sed -i '1d' 4_5.vcf\n",
    "echo \"POS,REF,ALT,QUAL,DEPTH,REF_READS,ALT_READS,TYPE,GENE\" > \"$s\"_SNPs_5.vcf\n",
    "    while IFS= read -r line; \n",
    "    do\n",
    "    pos=$(echo \"\"$line\"\" | sed  's/,.*$//');\n",
    "    nc_region='TRUE'\n",
    "        for c in \"${!gene_csv[@]}\"\n",
    "        do\n",
    "            if (( \"$pos\" >= \"${start_csv[$c]}\" && \"$pos\" <= \"${end_csv[$c]}\" ));\n",
    "            then\n",
    "                nc_region='FALSE'\n",
    "                echo \"\"$line\",\"\"${gene_csv[$c]}\"\",\" >> \"$s\"_SNPs_5.vcf;\n",
    "            fi\n",
    "        done\n",
    "        if [[ \"$nc_region\" == 'TRUE' ]];\n",
    "        then \n",
    "            echo \"\"$line\",-,\" >> \"$s\"_SNPs_5.vcf;\n",
    "        fi\n",
    "        nc_region='TRUE'\n",
    "    done < 4_5.vcf\n",
    "rm 1_5.vcf\n",
    "rm 2_5.vcf\n",
    "rm 3_5.vcf\n",
    "rm 4_5.vcf\n",
    "done\n",
    "\n",
    "#VCF Formatting for 1% SNP read support\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "cd \"$snp_dir\"/\"$s\"/\"$s\"_Export/\n",
    "sed '1,60d' \"$snp_dir\"/\"$s\"/\"$s\"_Export/bwa_paired_\"$s\"_sorted_freebayes_0.01_SNPs.vcf > 1_1.vcf\n",
    "sed -e \"s/AB\\=.*:GL//g\" -e \"s/.\\/.\\://g\" -e \"s/\\:/\\,/1\" -e \"s/\\:.*//g\" -e \"s/\\s\\+/,/g\" -e \"s/\\,ID//g\" -e \"s/\\,FILTER\\,INFO\\,FORMAT\\,unknown/\\,DEPTH\\,REF\\_READS\\,ALT\\_READS\\,TYPE/g\" -e \"s/\\,\\.//g\" -e \"s/\\#CHROM\\,//g\" -e \"s/\\,/\\:/1\" -e \"s/.*\\://g\" -e \"s/REF/POS\\,REF/g\" 1_1.vcf > 2_1.vcf\n",
    "echo \"POS,REF,ALT,QUAL,DEPTH,REF_READS,ALT_READS,TYPE,GENE\" > 3_1.vcf\n",
    "    while IFS= read -r line; \n",
    "    do\n",
    "        if [[ \"$line\" =~ \",A,\" ]] && [[ \"$line\" =~ \",C,\" ]]\n",
    "        then\n",
    "           echo \"\"$line\",transversion\" >> 3_1.vcf\n",
    "        elif [[ \"$line\" =~ \",A,\" ]] && [[ \"$line\" =~ \",G,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transition\" >> 3_1.vcf\n",
    "        elif [[ \"$line\" =~ \",A,\" ]] && [[ \"$line\" =~ \",T,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transversion\" >> 3_1.vcf\n",
    "        elif [[ \"$line\" =~ \",C,\" ]] && [[ \"$line\" =~ \",G,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transversion\" >> 3_1.vcf\n",
    "        elif [[ \"$line\" =~ \",C,\" ]] && [[ \"$line\" =~ \",T,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transition\" >> 3_1.vcf\n",
    "        elif [[ \"$line\" =~ \",G,\" ]] && [[ \"$line\" =~ \",T,\" ]]\n",
    "        then\n",
    "            echo \"\"$line\",transversion\" >> 3_1.vcf\n",
    "        else\n",
    "            echo \"\"$line\",substitution\" >> 3_1.vcf    \n",
    "        fi\n",
    "    done < 2_1.vcf\n",
    "sed '2d' 3_1.vcf  > 4_1.vcf\n",
    "sed -i '1d' 4_1.vcf\n",
    "echo \"POS,REF,ALT,QUAL,DEPTH,REF_READS,ALT_READS,TYPE,GENE\" > \"$s\"_SNPs_1.vcf\n",
    "    while IFS= read -r line; \n",
    "    do\n",
    "    pos=$(echo \"\"$line\"\" | sed  's/,.*$//');\n",
    "    nc_region='TRUE'\n",
    "        for c in \"${!gene_csv[@]}\"\n",
    "        do\n",
    "            if (( \"$pos\" >= \"${start_csv[$c]}\" && \"$pos\" <= \"${end_csv[$c]}\" ));\n",
    "            then\n",
    "                nc_region='FALSE'\n",
    "                echo \"\"$line\",\"\"${gene_csv[$c]}\"\",\" >> \"$s\"_SNPs_1.vcf;\n",
    "            fi\n",
    "        done\n",
    "        if [[ \"$nc_region\" == 'TRUE' ]];\n",
    "        then \n",
    "            echo \"\"$line\",-,\" >> \"$s\"_SNPs_1.vcf;\n",
    "        fi\n",
    "        nc_region='TRUE'\n",
    "    done < 4_1.vcf\n",
    "rm 1_1.vcf\n",
    "rm 2_1.vcf\n",
    "rm 3_1.vcf\n",
    "rm 4_1.vcf\n",
    "done\n",
    "\n",
    "#Concatenating the VCFs for 50% SNP read support\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "cd \"$snp_dir\"/\"$s\"/\"$s\"_Export/ \n",
    "echo \"POS,REF,ALT,QUAL,DEPTH,REF_READS,ALT_READS,TYPE,GENE, Start, End, Possible SNP Start (Direct), Possible SNP End (Direct), Possible SNP Start (Complement), Possible SNP End (Complement), AA Change\" > 1_50.csv\n",
    "sed '1d' \"$analysis_dir\"/results/\"$s\"/aa_changes/\"$s\"_aa_change_50.csv > working_50.csv\n",
    "sed '1d' \"$snp_dir\"/\"$s\"/\"$s\"_Export/\"$s\"_SNPs_50.vcf > working_50.vcf\n",
    "st_dir=( $(tail -n +1 working_50.csv  | cut -d ',' -f4) )\n",
    "end_dir=( $(tail -n +1 working_50.csv  | cut -d ',' -f5) )\n",
    "st_cmp=( $(tail -n +1 working_50.csv  | cut -d ',' -f6) )\n",
    "end_cmp=( $(tail -n +1 working_50.csv  | cut -d ',' -f7) )\n",
    "gene_csv=( $(tail -n +1 working_50.csv  | cut -d ',' -f1) )\n",
    "wc -l < working_50.csv\n",
    "    while IFS= read -r line1; \n",
    "    do\n",
    "    lineadder='TRUE'\n",
    "    pos=( $(echo \"\"$line1\"\" | sed  's/,.*$//') )\n",
    "    pre_gene=( $(echo \"\"$line1\"\" | sed 's/,\\([^,]*\\)$/\\1/') )\n",
    "    gene_vcf=( $(echo \"\"$pre_gene\"\" | sed 's/^.*,//') )\n",
    "        for c in \"${!st_dir[@]}\"\n",
    "        do\n",
    "            if (( \"$pos\" >= \"${st_dir[$c]}\" && \"$pos\" <= \"${end_dir[$c]}\" )) && [ ${gene_csv[$c]} = $gene_vcf ];\n",
    "            then\n",
    "                lineadder='FALSE';\n",
    "                line2=$(grep -m 1 \"\"${st_dir[$c]}\",\"${end_dir[$c]}\",\" working_50.csv | head -n1)\n",
    "                echo \"\"$line1\"\"$line2\"\" >> 1_50.csv;\n",
    "                sed -i '0,/'\"${st_dir[$c]}\"','\"${end_dir[$c]}\"'/{//d}' working_50.csv \n",
    "                sed -i '0,/'\"\"$pos\"\"'/{//d}' working_50.vcf\n",
    "                unset st_dir[$c]\n",
    "                unset end_dir[$c]\n",
    "                unset st_cmp[$c]\n",
    "                unset end_cmp[$c]\n",
    "                unset gene_csv[$c]\n",
    "                break\n",
    "            elif (( \"$pos\" >= \"${st_cmp[$c]}\" && \"$pos\" <= \"${end_cmp[$c]}\" )) && [ ${gene_csv[$c]} = $gene_vcf ];\n",
    "            then\n",
    "                lineadder='FALSE';\n",
    "                line2=$(grep -m 1 \"\"${st_cmp[$c]}\",\"${end_cmp[$c]}\",\" working_50.csv | head -n1)\n",
    "                echo \"\"$line1\"\"$line2\"\" >> 1_50.csv;\n",
    "                sed -i '0,/'\"${st_cmp[$c]}\"','\"${end_cmp[$c]}\"'/{//d}' working_50.csv \n",
    "                sed -i '0,/'\"\"$pos\"\"'/{//d}' working_50.vcf\n",
    "                unset st_dir[$c]\n",
    "                unset end_dir[$c]\n",
    "                unset st_cmp[$c]\n",
    "                unset end_cmp[$c]\n",
    "                unset gene_csv[$c]\n",
    "                break\n",
    "            fi\n",
    "        done\n",
    "        if [[ \"$lineadder\" == 'TRUE' ]];\n",
    "        then\n",
    "            echo \"\"$line1\"-,-,-,-,-,-,-,-\" >> 1_50.csv;\n",
    "            sed -i '/'\"$pos\"'/d' working_50.vcf\n",
    "        fi\n",
    "    done < working_50.vcf\n",
    "remove='10,11,12,13,14,15,16'\n",
    "cut -d ',' -f$remove --complement 1_50.csv > 2_50.csv\n",
    "awk -F, '{ print $1 \",\" $9 \",\" $2 \",\" $3 \",\" $4 \",\" $5 \",\" $6 \",\" $7 \",\" $8 \",\" $10}' 2_50.csv > 3_50.csv\n",
    "awk -F, '{ print $1 \",\" $2 \",\" $6 \",\" $3 \",\" $7 \",\" $4 \",\" $8 \",\" $5 \",\" $9 \",\" $10}' 3_50.csv > \"$s\"_SNPs_50_final.csv\n",
    "cp \"$s\"_SNPs_50_final.csv \"$analysis_dir\"/results/\"$s\"/\n",
    "rm working_50.vcf\n",
    "rm working_50.csv\n",
    "rm 1_50.csv\n",
    "rm 2_50.csv\n",
    "rm 3_50.csv\n",
    "done\n",
    "\n",
    "#Concatenating the VCFs for 10% SNP read support\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "cd \"$snp_dir\"/\"$s\"/\"$s\"_Export/ \n",
    "echo \"POS,REF,ALT,QUAL,DEPTH,REF_READS,ALT_READS,TYPE,GENE, Start, End, Possible SNP Start (Direct), Possible SNP End (Direct), Possible SNP Start (Complement), Possible SNP End (Complement), AA Change\" > 1_10.csv\n",
    "sed '1d' \"$analysis_dir\"/results/\"$s\"/aa_changes/\"$s\"_aa_change_10.csv > working_10.csv\n",
    "sed '1d' \"$snp_dir\"/\"$s\"/\"$s\"_Export/\"$s\"_SNPs_10.vcf > working_10.vcf\n",
    "st_dir=( $(tail -n +1 working_10.csv  | cut -d ',' -f4) )\n",
    "end_dir=( $(tail -n +1 working_10.csv  | cut -d ',' -f5) )\n",
    "st_cmp=( $(tail -n +1 working_10.csv  | cut -d ',' -f6) )\n",
    "end_cmp=( $(tail -n +1 working_10.csv  | cut -d ',' -f7) )\n",
    "gene_csv=( $(tail -n +1 working_10.csv  | cut -d ',' -f1) )\n",
    "wc -l < working_10.csv\n",
    "    while IFS= read -r line1; \n",
    "    do\n",
    "    lineadder='TRUE'\n",
    "    pos=( $(echo \"\"$line1\"\" | sed  's/,.*$//') )\n",
    "    pre_gene=( $(echo \"\"$line1\"\" | sed 's/,\\([^,]*\\)$/\\1/') )\n",
    "    gene_vcf=( $(echo \"\"$pre_gene\"\" | sed 's/^.*,//') )\n",
    "        for c in \"${!st_dir[@]}\"\n",
    "        do\n",
    "            if (( \"$pos\" >= \"${st_dir[$c]}\" && \"$pos\" <= \"${end_dir[$c]}\" )) && [ \"${gene_csv[$c]}\" = \"$gene_vcf\" ];\n",
    "            then\n",
    "                lineadder='FALSE';\n",
    "                line2=$(grep -m 1 \"\"${st_dir[$c]}\",\"${end_dir[$c]}\",\" working_10.csv | head -n1)\n",
    "                echo \"\"$line1\"\"$line2\"\" >> 1_10.csv;\n",
    "                sed -i '0,/'\"${st_dir[$c]}\"','\"${end_dir[$c]}\"'/{//d}' working_10.csv \n",
    "                sed -i '0,/'\"\"$pos\"\"'/{//d}' working_10.vcf\n",
    "                unset st_dir[$c]\n",
    "                unset end_dir[$c]\n",
    "                unset st_cmp[$c]\n",
    "                unset end_cmp[$c]\n",
    "                unset gene_csv[$c]\n",
    "                break\n",
    "            elif (( \"$pos\" >= \"${st_cmp[$c]}\" && \"$pos\" <= \"${end_cmp[$c]}\" )) && [ \"${gene_csv[$c]}\" = \"$gene_vcf\" ];\n",
    "            then\n",
    "                lineadder='FALSE';\n",
    "                line2=$(grep -m 1 \"\"${st_cmp[$c]}\",\"${end_cmp[$c]}\",\" working_10.csv | head -n1)\n",
    "                echo \"\"$line1\"\"$line2\"\" >> 1_10.csv;\n",
    "                sed -i '0,/'\"${st_cmp[$c]}\"','\"${end_cmp[$c]}\"'/{//d}' working_10.csv \n",
    "                sed -i '0,/'\"\"$pos\"\"'/{//d}' working_10.vcf\n",
    "                unset st_dir[$c]\n",
    "                unset end_dir[$c]\n",
    "                unset st_cmp[$c]\n",
    "                unset end_cmp[$c]\n",
    "                unset gene_csv[$c]\n",
    "                break\n",
    "            fi\n",
    "        done\n",
    "        if [[ \"$lineadder\" == 'TRUE' ]];\n",
    "        then\n",
    "            echo \"\"$line1\"-,-,-,-,-,-,-,-\" >> 1_10.csv;\n",
    "            sed -i '/'\"$pos\"'/d' working_10.vcf\n",
    "        fi\n",
    "    done < working_10.vcf\n",
    "remove='10,11,12,13,14,15,16'\n",
    "cut -d ',' -f$remove --complement 1_10.csv > 2_10.csv\n",
    "awk -F, '{ print $1 \",\" $9 \",\" $2 \",\" $3 \",\" $4 \",\" $5 \",\" $6 \",\" $7 \",\" $8 \",\" $10}' 2_10.csv > 3_10.csv\n",
    "awk -F, '{ print $1 \",\" $2 \",\" $6 \",\" $3 \",\" $7 \",\" $4 \",\" $8 \",\" $5 \",\" $9 \",\" $10}' 3_10.csv > \"$s\"_SNPs_10_final.csv\n",
    "cp \"$s\"_SNPs_10_final.csv \"$analysis_dir\"/results/\"$s\"/\n",
    "rm working_10.vcf\n",
    "rm working_10.csv\n",
    "rm 1_10.csv\n",
    "rm 2_10.csv\n",
    "rm 3_10.csv\n",
    "done\n",
    "\n",
    "#Concatenating the VCFs for 5% SNP read support\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "cd \"$snp_dir\"/\"$s\"/\"$s\"_Export/ \n",
    "echo \"POS,REF,ALT,QUAL,DEPTH,REF_READS,ALT_READS,TYPE,GENE, Start, End, Possible SNP Start (Direct), Possible SNP End (Direct), Possible SNP Start (Complement), Possible SNP End (Complement), AA Change\" > 1_5.csv\n",
    "sed '1d' \"$analysis_dir\"/results/\"$s\"/aa_changes/\"$s\"_aa_change_5.csv > working_5.csv\n",
    "sed '1d' \"$snp_dir\"/\"$s\"/\"$s\"_Export/\"$s\"_SNPs_5.vcf > working_5.vcf\n",
    "st_dir=( $(tail -n +1 working_5.csv  | cut -d ',' -f4) )\n",
    "end_dir=( $(tail -n +1 working_5.csv  | cut -d ',' -f5) )\n",
    "st_cmp=( $(tail -n +1 working_5.csv  | cut -d ',' -f6) )\n",
    "end_cmp=( $(tail -n +1 working_5.csv  | cut -d ',' -f7) )\n",
    "gene_csv=( $(tail -n +1 working_5.csv  | cut -d ',' -f1) )\n",
    "wc -l < working_5.csv\n",
    "    while IFS= read -r line1; \n",
    "    do\n",
    "    lineadder='TRUE'\n",
    "    pos=( $(echo \"\"$line1\"\" | sed  's/,.*$//') )\n",
    "    pre_gene=( $(echo \"\"$line1\"\" | sed 's/,\\([^,]*\\)$/\\1/') )\n",
    "    gene_vcf=( $(echo \"\"$pre_gene\"\" | sed 's/^.*,//') )\n",
    "        for c in \"${!st_dir[@]}\"\n",
    "        do\n",
    "            if (( \"$pos\" >= \"${st_dir[$c]}\" && \"$pos\" <= \"${end_dir[$c]}\" )) && [ ${gene_csv[$c]} = $gene_vcf ];\n",
    "            then\n",
    "                lineadder='FALSE';\n",
    "                line2=$(grep -m 1 \"\"${st_dir[$c]}\",\"${end_dir[$c]}\",\" working_5.csv | head -n1)\n",
    "                echo \"\"$line1\"\"$line2\"\" >> 1_5.csv;\n",
    "                sed -i '0,/'\"${st_dir[$c]}\"','\"${end_dir[$c]}\"'/{//d}' working_5.csv \n",
    "                sed -i '0,/'\"\"$pos\"\"'/{//d}' working_5.vcf\n",
    "                unset st_dir[$c]\n",
    "                unset end_dir[$c]\n",
    "                unset st_cmp[$c]\n",
    "                unset end_cmp[$c]\n",
    "                unset gene_csv[$c]\n",
    "                break\n",
    "            elif (( \"$pos\" >= \"${st_cmp[$c]}\" && \"$pos\" <= \"${end_cmp[$c]}\" )) && [ ${gene_csv[$c]} = $gene_vcf ];\n",
    "            then\n",
    "                lineadder='FALSE';\n",
    "                line2=$(grep -m 1 \"\"${st_cmp[$c]}\",\"${end_cmp[$c]}\",\" working_5.csv | head -n1)\n",
    "                echo \"\"$line1\"\"$line2\"\" >> 1_5.csv;\n",
    "                sed -i '0,/'\"${st_cmp[$c]}\"','\"${end_cmp[$c]}\"'/{//d}' working_5.csv \n",
    "                sed -i '0,/'\"\"$pos\"\"'/{//d}' working_5.vcf\n",
    "                unset st_dir[$c]\n",
    "                unset end_dir[$c]\n",
    "                unset st_cmp[$c]\n",
    "                unset end_cmp[$c]\n",
    "                unset gene_csv[$c]\n",
    "                break\n",
    "            fi\n",
    "        done\n",
    "        if [[ \"$lineadder\" == 'TRUE' ]];\n",
    "        then\n",
    "            echo \"\"$line1\"-,-,-,-,-,-,-,-\" >> 1_5.csv;\n",
    "            sed -i '/'\"$pos\"'/d' working_5.vcf\n",
    "        fi\n",
    "    done < working_5.vcf\n",
    "remove='10,11,12,13,14,15,16'\n",
    "cut -d ',' -f$remove --complement 1_5.csv > 2_5.csv\n",
    "awk -F, '{ print $1 \",\" $9 \",\" $2 \",\" $3 \",\" $4 \",\" $5 \",\" $6 \",\" $7 \",\" $8 \",\" $10}' 2_5.csv > 3_5.csv\n",
    "awk -F, '{ print $1 \",\" $2 \",\" $6 \",\" $3 \",\" $7 \",\" $4 \",\" $8 \",\" $5 \",\" $9 \",\" $10}' 3_5.csv > \"$s\"_SNPs_5_final.csv\n",
    "cp \"$s\"_SNPs_5_final.csv \"$analysis_dir\"/results/\"$s\"/\n",
    "rm working_5.vcf\n",
    "rm working_5.csv\n",
    "rm 1_5.csv\n",
    "rm 2_5.csv\n",
    "rm 3_5.csv\n",
    "done\n",
    "\n",
    "#Concatenating the VCFs for 1% SNP read support\n",
    "for s in \"${samples[@]}\"\n",
    "do\n",
    "cd \"$snp_dir\"/\"$s\"/\"$s\"_Export/ \n",
    "echo \"POS,REF,ALT,QUAL,DEPTH,REF_READS,ALT_READS,TYPE,GENE, Start, End, Possible SNP Start (Direct), Possible SNP End (Direct), Possible SNP Start (Complement), Possible SNP End (Complement), AA Change\" > 1_1.csv\n",
    "sed '1d' \"$analysis_dir\"/results/\"$s\"/aa_changes/\"$s\"_aa_change_1.csv > working_1.csv\n",
    "sed '1d' \"$snp_dir\"/\"$s\"/\"$s\"_Export/\"$s\"_SNPs_1.vcf > working_1.vcf\n",
    "st_dir=( $(tail -n +1 working_1.csv  | cut -d ',' -f4) )\n",
    "end_dir=( $(tail -n +1 working_1.csv  | cut -d ',' -f5) )\n",
    "st_cmp=( $(tail -n +1 working_1.csv  | cut -d ',' -f6) )\n",
    "end_cmp=( $(tail -n +1 working_1.csv  | cut -d ',' -f7) )\n",
    "gene_csv=( $(tail -n +1 working_1.csv  | cut -d ',' -f1) )\n",
    "wc -l < working_1.csv\n",
    "    while IFS= read -r line1; \n",
    "    do\n",
    "    lineadder='TRUE'\n",
    "    pos=( $(echo \"\"$line1\"\" | sed  's/,.*$//') )\n",
    "    pre_gene=( $(echo \"\"$line1\"\" | sed 's/,\\([^,]*\\)$/\\1/') )\n",
    "    gene_vcf=( $(echo \"\"$pre_gene\"\" | sed 's/^.*,//') )\n",
    "        for c in \"${!st_dir[@]}\"\n",
    "        do\n",
    "            if (( \"$pos\" >= \"${st_dir[$c]}\" && \"$pos\" <= \"${end_dir[$c]}\" )) && [ \"${gene_csv[$c]}\" = \"$gene_vcf\" ];\n",
    "            then\n",
    "                lineadder='FALSE';\n",
    "                line2=$(grep -m 1 \"\"${st_dir[$c]}\",\"${end_dir[$c]}\",\" working_1.csv | head -n1)\n",
    "                echo \"\"$line1\"\"$line2\"\" >> 1_1.csv;\n",
    "                sed -i '0,/'\"${st_dir[$c]}\"','\"${end_dir[$c]}\"'/{//d}' working_1.csv \n",
    "                sed -i '0,/'\"\"$pos\"\"'/{//d}' working_1.vcf\n",
    "                unset st_dir[$c]\n",
    "                unset end_dir[$c]\n",
    "                unset st_cmp[$c]\n",
    "                unset end_cmp[$c]\n",
    "                unset gene_csv[$c]\n",
    "                break\n",
    "            elif (( \"$pos\" >= \"${st_cmp[$c]}\" && \"$pos\" <= \"${end_cmp[$c]}\" )) && [ \"${gene_csv[$c]}\" = \"$gene_vcf\" ];\n",
    "            then\n",
    "                lineadder='FALSE';\n",
    "                line2=$(grep -m 1 \"\"${st_cmp[$c]}\",\"${end_cmp[$c]}\",\" working_1.csv | head -n1)\n",
    "                echo \"\"$line1\"\"$line2\"\" >> 1_1.csv;\n",
    "                sed -i '0,/'\"${st_cmp[$c]}\"','\"${end_cmp[$c]}\"'/{//d}' working_1.csv \n",
    "                sed -i '0,/'\"\"$pos\"\"'/{//d}' working_1.vcf\n",
    "                unset st_dir[$c]\n",
    "                unset end_dir[$c]\n",
    "                unset st_cmp[$c]\n",
    "                unset end_cmp[$c]\n",
    "                unset gene_csv[$c]\n",
    "                break\n",
    "            fi\n",
    "        done\n",
    "        if [[ \"$lineadder\" == 'TRUE' ]];\n",
    "        then\n",
    "            echo \"\"$line1\"-,-,-,-,-,-,-,-\" >> 1_1.csv;\n",
    "            sed -i '/'\"$pos\"'/d' working_1.vcf\n",
    "        fi\n",
    "    done < working_1.vcf\n",
    "remove='10,11,12,13,14,15,16'\n",
    "cut -d ',' -f$remove --complement 1_1.csv > 2_1.csv\n",
    "awk -F, '{ print $1 \",\" $9 \",\" $2 \",\" $3 \",\" $4 \",\" $5 \",\" $6 \",\" $7 \",\" $8 \",\" $10}' 2_1.csv > 3_1.csv\n",
    "awk -F, '{ print $1 \",\" $2 \",\" $6 \",\" $3 \",\" $7 \",\" $4 \",\" $8 \",\" $5 \",\" $9 \",\" $10}' 3_1.csv > \"$s\"_SNPs_1_final.csv\n",
    "cp \"$s\"_SNPs_1_final.csv \"$analysis_dir\"/results/\"$s\"/\n",
    "rm working_1.vcf\n",
    "rm working_1.csv\n",
    "rm 1_1.csv\n",
    "rm 2_1.csv\n",
    "rm 3_1.csv\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b7ef3e8-9094-4908-880f-7cca8c9224a6",
   "metadata": {},
   "source": [
    "## Protein Function Prediction\n",
    "Predicting possible alterations in protein function caused by AA changes. \n",
    "#### Provean\n",
    "For more information on the command line PROVEAN tool see [here](http://provean.jcvi.org/downloads/README). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9df2c7fe-e5b5-43b3-9d6f-5a82416dd620",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#PROVEAN predictions for 50% SNP read support\n",
    "#for s in \"${samples[@]}\"\n",
    "#do\n",
    "#cd \"$output_dir\"/\"$s\"_50/aa_changes/\n",
    "#mkdir PROVEAN\n",
    "#cd \"$output_dir\"/\"$s\"_50/aa_changes/PROVEAN/\n",
    "#    for gene in \"${gene_csv[@]}\"\n",
    "#    do\n",
    "#    /home/fuberlin/Programs/provean-1.1.5/scripts/provean.sh -q \"$output_dir\"/\"$refname\"/gene-\"$gene\"_ref.fasta -v \"$output_dir\"/\"$s\"_10/aa_changes/\"$gene\".var > \"$gene\"_PROVEAN.txt\n",
    "#    done\n",
    "#done\n",
    "#create a csv with the change, score and classification (if statement with neutral and deleterious)\n",
    "#use csvtool to append the scores and classification of the aa changes into the main csv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0675329-1c06-47b9-88c6-d754f880d591",
   "metadata": {},
   "source": [
    "Deleting unnecessary files and folders:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "636660e0-0b41-4f45-93f6-034b15a45485",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Housekeeping, removing all unecessary files\n",
    "\n",
    "#rm -r \"$output_dir\""
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
